### ISSUE TITLE
Calling zip_entry_open after zip_close gives an ugly error

### ISSUE THREAD

**Issue by gynt** [kuba--/zip#209]
The error happens here: https://github.com/kuba--/zip/blob/master/src/zip.c#L891 and in my testing it happened also in the CLEANUP call in the line right after.
It happens because the pointer to the zip entry structure has become invalid after closing the zip, which is expected behaviour of course.

However, it would be more informational to make zip_entry_open return an error code that indicates that the zip was closed.

**Comment by codylico** [comment]
_outsider comment; feel free to ignore:_

> It happens because the pointer to the zip entry structure has become invalid after closing the zip, which is expected behaviour of course.

As you noted, the `CLEANUP` macro used in the `zip` library frees the pointer to the zip structure, entry and all. Once that free happens, the heap/operating system/memory allocator reclaims that memory (and possibly uses it for something else).

> However, it would be more informational to make zip_entry_open return an error code that indicates that the zip was closed.

At first glance, to implement a proper error code would mean that the `zip` library would have remember the pointers to the zip structures it closed and freed, then check that list whenever the user calls `zip_entry_open`, etc. I expect that the functions you mention would have the same problem if passed a junk pointer (such as `(void*)1234567`).

If you don't mind me asking, what kind of ugly error do you see?

**Comment by gynt** [comment]
Hmm, so I figured out my issue is actually quite fundamental. `zip_close` frees the complete zip_t struct with a `CLEANUP`, while I was expecting it to remember the zip was closed. So `zip_close` is more like a `zip_delete` or `zip_destroy`.

An example of a specific error I am getting is an access violation at the `CLEANUP` statement: Access violation reading location 0xDDDDDDD4. Basically the code is trying to free the memory for `zip->entry.name` which is invalid memory at that point.

This code reproduces the error:
```cpp
int main()
{
	zip_t* zip = zip_open("issue-209.zip", 0, 'r');

	if (zip_entry_open(zip, "miniz.h") != 0) {
		// Should not happen
	}

	zip_close(zip);

	// The error happens inside zip_entry_open
	return zip_entry_open(zip, "miniz.h");
}
```

**Comment by kuba--** [comment]
I wonder if we need to fix the issue, and check everywhere if archive is closed etc.
Mainly, because the problem is not really in library, but in the way how it's used.

Like if you take a look into stdlib, e.g.:
```c
int main() {                                                                    
  char *s;                                                                    
  strcpy(s, "abcdefgh");                                                      
} 
```
`strcpy` does not return an error, like _not allocated memory_ or _not enough space_, the code will just crash.
In other words, in such a way we can break any code. You can't use variable after you free memory for it.

**Comment by gynt** [comment]
Yes you are right. The subtlety is in the fact `zip_open` returns a pointer, instead of a `zip_t` struct.
So what I am trying to say is that `zip_close` and `zip_open` sound like functions that operate on a struct in memory and open a stream.
As if:
```cpp
zip_t zip;
zip_open(&zip, "filename.zip", 0, "r")
```
But instead, the functions allocate and free memory. 

I fully admit that the error is very obvious and I actually also don't have good alternative suggestions:
`zip_create` and `zip_destroy` sound as if they are filesystem functions. `zip_alloc` and `zip_free` comes a bit closer, but is also kind of awkward.

An alternative is to separate the alloc and free logic from the open and close calls. Which would allow this pattern:
For example:
```cpp
zip_t zip;

if(!zip_open(&zip, "filename.zip", 0, "r")) {
  // catch the error
}
zip_close(&zip);

if(!zip_open(&zip, "another-filename.zip", 0, "r")) {
  // catch the error
}
zip_close(&zip);
```

These are just suggestions. It is not a very bad issue either. It does not need to be fixed, but perhaps clearly documented.

**Comment by kuba--** [comment]
Are you talking about naming conventions? `zip_open`/`zip_close` mainly opens/closes files, flushes central directory etc. and of course release resources.
Personally I don't see the point to change API (right now), but of course any pull request where you can set own allocator/deallocator is welcome.

**Comment by kuba--** [comment]
Feel free to reopen the issue, if it's needed.
